### PR TITLE
Removed static variables from F_Back code

### PR DESCRIPTION
**Pull request by FrostyMike**

### CODE REVIEW

File: C-BackEnd/src/xcodeml/c/decompile/XcXmpFactory.java [hunk 1]
```diff
@@ -59,10 +59,10 @@
      *
      * @return the name of temporary variable name.
      */
-    private static String _createTmpValueName()
+    /*private static String _createTmpValueName()
```
Comment (h-murai) [inline C-BackEnd/src/xcodeml/c/decompile/XcXmpFactory.java]: Is it possible to keep this function for future use?

Comment (FrostyMike) [reply C-BackEnd/src/xcodeml/c/decompile/XcXmpFactory.java]: Yes, I can revert this change.
Please note however, that currently it is not called anywhere in xcodeml-tools, so it's probably dead code.

File: C-BackEnd/src/xcodeml/c/decompile/XcProgramObj.java [hunk 2]
```diff
@@ -34,8 +36,16 @@
...
     private XcFuncDefObj _xmpMain = null;
 
+    private final IXmOption xmOption;
+
+    @Deprecated
     public XcProgramObj()
     {
+        xmOption = new XmOptionStatic();
+    }
+    public XcProgramObj(IXmOption xmOption)
```
Comment (clementval) [inline C-BackEnd/src/xcodeml/c/decompile/XcProgramObj.java:47]: NIT: empty line between ctors

File: C-BackEnd/src/xcodeml/c/decompile/XcXmpFactory.java [hunk 3]
```diff
@@ -22,7 +22,7 @@
  */
 public final class XcXmpFactory
 {
-    private static int _tmpVarId = 0;
+    //private static int _tmpVarId = 0;
```
Comment (clementval) [inline C-BackEnd/src/xcodeml/c/decompile/XcXmpFactory.java]: Should be removed?

**Review by clementval — COMMENTED** [review]
Just for my understanding, why is this necessary. Maybe some information in the PR description would be nice for future reference.

Same as #174, the PR title is a bit misleading. It is not only removing static variables but also adding some new classes. Should it be split in two PR? 

@h-murai As for #174 I let you give the final approval.

**Comment by FrostyMike** [comment]
> Just for my understanding, why is this necessary.

It allows F_Back to be used as a library, instead of executable, in a (thread-)safe manner. This is beneficial for new CLAW driver, 
other users probably won't see any difference.

> Same as #174, the PR title is a bit misleading. It is not only removing static variables but also adding some new classes. Should it be split in two PR?

No, new classes are added to make removal of static variables possible. Separation is not advisable.

**Comment by clementval** [comment]
> No, new classes are added to make removal of static variables possible. Separation is not advisable.

Anyway a bit more information in the description wouldn't hurt.

File: C-BackEnd/src/xcodeml/c/decompile/XcXmpFactory.java (hunk 1 above)
Comment (h-murai) [reply C-BackEnd/src/xcodeml/c/decompile/XcXmpFactory.java]: Ok. I'll restore it if we need it in the future.

**Comment by h-murai** [comment]
@FrostyMike Please follow the comments from @clementval , and then I'll approve this PR.

File: C-BackEnd/src/xcodeml/c/decompile/XcProgramObj.java (hunk 2 above)
Comment (FrostyMike) [reply C-BackEnd/src/xcodeml/c/decompile/XcProgramObj.java:47]: Fixed

File: C-BackEnd/src/xcodeml/c/decompile/XcXmpFactory.java (hunk 3 above)
Comment (FrostyMike) [reply C-BackEnd/src/xcodeml/c/decompile/XcXmpFactory.java]: Fixed

**Comment by FrostyMike** [comment]
> > No, new classes are added to make removal of static variables possible. Separation is not advisable.
> 
> Anyway a bit more information in the description wouldn't hurt.

Description: refactored XmOption class, which is a collection of static libraries. It was replaced by non-static XmOptionLocal and static XmOptionStatic (kept for support of legacy usages), both accessed via new interface IXmOption. Replaced XmOption with these 2 classes in all usages in this repo.

**Comment by FrostyMike** [comment]
> @FrostyMike Please follow the comments from @clementval , and then I'll approve this PR.

Done

**Pull request merged**